### PR TITLE
Refactor multipart download to a more async model

### PR DESCRIPTION
**Pull request by andrross**
The previous approach of kicking off the stream requests for all parts
of a file did not work well for very large files. For example, a 20GiB
file uploaded in 16MiB parts will consist of 1200+ parts. When we
attempted to initiate streaming for all parts concurrently, some parts
would hit a client timeout after 2 minutes without being able to get a
connection due to the other parts not having been completed in that time
frame. This refactoring adds yet another layer of indirection in order
to allow the code that is actually writing the destination file to
control the rate at which streams are started. This should allow for
downloading files consisting of arbitrarily many parts at any connection
speed.

This commit also wires in the download rate limiter so that the
`indices.recovery.max_bytes_per_sec` is properly honored.

This PR supersedes #10284

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff
- [x] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))
- [x] GitHub issue/PR created in [OpenSearch documentation repo](https://github.com/opensearch-project/documentation-website) for the required public documentation changes (#[Issue/PR number])

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by kotwanikunal** [comment]
@andrross I think you need to rebase. There is a serialization version change which might be causing test failures.

**Review by kotwanikunal — APPROVED** [review]

File: server/src/main/java/org/opensearch/common/blobstore/stream/read/listener/ReadContextListener.java [hunk 1]
```diff
@@ -10,56 +10,155 @@
...
+            fileLocation,
+            groupedListener,
+            threadPool.executor(ThreadPool.Names.REMOTE_RECOVERY),
+            rateLimiter
+        );
+        // We have a dedicated thread pool for draining streams and writing them to a file on the
+        // local disk. However, we want more concurrent requests to fetch stream parts to be
+        // happening asynchronously so that there is always a queue of streams ready to be consumed
+        // by the REMOTE_RECOVERY thread pool.
+        final int concurrentStreamRequests = 20;
```
Comment (andrross) [inline server/src/main/java/org/opensearch/common/blobstore/stream/read/listener/ReadContextListener.java:79]: This constant isn't ideal. In my limiting testing this value seems to work pretty well. I'm not exactly sure how to parameterize or scale this value, so just starting simple with a constant value. The underlying connection pool for the S3 repository has a default size of 100 (also not scaled by host size).

File: plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/S3BlobContainer.java [hunk 2]
```diff
@@ -241,37 +241,23 @@ public void readBlobAsync(String blobName, ActionListener<ReadContext> listener)
...
+                    blobPartInputStreamFutures.add(() -> getBlobPartInputStreamContainer(s3AsyncClient, bucketName, blobKey, null));
                 } else {
                     // S3 multipart files use 1 to n indexing
                     for (int partNumber = 1; partNumber <= numberOfParts; partNumber++) {
-                        blobPartInputStreamFutures.add(getBlobPartInputStreamContainer(s3AsyncClient, bucketName, blobKey, partNumber));
+                        final int innerPartNumber = partNumber;
+                        blobPartInputStreamFutures.add(
+                            () -> getBlobPartInputStreamContainer(s3AsyncClient, bucketName, blobKey, innerPartNumber)
+                        );
                     }
```
Comment (gbbafna) [inline plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/S3BlobContainer.java:258]: can we please test with this with restoring more number of shards at once ? The potential problem i see with this approach is the high number of connections we are opening . If we open it across shards and they get rate limited/thread pool starved for some time, the connections pool would get exhausted.

Comment (gbbafna) [reply plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/S3BlobContainer.java:258]: So when does connection get opened exactly ? I might be wrong here if we open connections lazily , 20 at a time max per file.

File: server/src/main/java/org/opensearch/common/blobstore/stream/read/listener/ReadContextListener.java (hunk 1 above)
Comment (gbbafna) [reply server/src/main/java/org/opensearch/common/blobstore/stream/read/listener/ReadContextListener.java:79]: can we make this dynamic with default size of 20 ?

File: plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/S3BlobContainer.java (hunk 2 above)
Comment (andrross) [reply plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/S3BlobContainer.java:258]: It is 20 per file, but we restore one file at a time. So in effect it is 20 at a time per shard.

File: server/src/main/java/org/opensearch/common/blobstore/stream/read/listener/ReadContextListener.java (hunk 1 above)
Comment (Bukhtawar) [reply server/src/main/java/org/opensearch/common/blobstore/stream/read/listener/ReadContextListener.java:79]: +1 on making this tunable

File: server/src/main/java/org/opensearch/common/blobstore/stream/read/listener/FilePartWriter.java [hunk 3]
```diff
@@ -8,83 +8,37 @@
...
-                        streamOffset += bytesRead;
-                    }
+    public static void write(Path fileLocation, InputStreamContainer stream, UnaryOperator<InputStream> rateLimiter) throws IOException {
+        try (FileChannel outputFileChannel = FileChannel.open(fileLocation, StandardOpenOption.WRITE, StandardOpenOption.CREATE)) {
+            try (InputStream inputStream = rateLimiter.apply(stream.getInputStream())) {
+                long streamOffset = stream.getOffset();
+                final byte[] buffer = new byte[BUFFER_SIZE];
+                int bytesRead;
+                while ((bytesRead = inputStream.read(buffer)) != -1) {
+                    Channels.writeToChannel(buffer, 0, bytesRead, outputFileChannel, streamOffset);
```
[the remark this answers is not in the document]
Comment (andrross) [reply server/src/main/java/org/opensearch/common/blobstore/stream/read/listener/FilePartWriter.java:38]: Ack, we should roll that into #10347 and do the fsync upon completion of the last part before copying to the destination file name.

File: server/src/main/java/org/opensearch/common/blobstore/stream/read/listener/ReadContextListener.java [hunk 4]
```diff
@@ -10,56 +10,155 @@
...
+    private final UnaryOperator<InputStream> rateLimiter;
 
-    public ReadContextListener(String fileName, Path fileLocation, ThreadPool threadPool, ActionListener<String> completionListener) {
-        this.fileName = fileName;
+    public ReadContextListener(
+        String blobName,
+        Path fileLocation,
+        ActionListener<String> completionListener,
+        ThreadPool threadPool,
+        UnaryOperator<InputStream> rateLimter
```
Comment (kotwanikunal) [inline server/src/main/java/org/opensearch/common/blobstore/stream/read/listener/ReadContextListener.java:48]: nit: typo in rateLimiter : `UnaryOperator<InputStream> rateLimiter`

File: server/src/main/java/org/opensearch/common/blobstore/stream/read/listener/ReadContextListener.java [hunk 5]
```diff
@@ -10,56 +10,155 @@
...
+        private final Path fileLocation;
+        private final GroupedActionListener<String> completionListener;
+        private final Executor executor;
+        private final UnaryOperator<InputStream> rateLimiter;
+
+        private StreamPartProcessor(
+            Queue<ReadContext.StreamPartCreator> queue,
+            AtomicBoolean anyPartStreamFailed,
+            Path fileLocation,
+            GroupedActionListener<String> completionListener,
```
Comment (kotwanikunal) [inline server/src/main/java/org/opensearch/common/blobstore/stream/read/listener/ReadContextListener.java:105]:
```suggestion
            ActionListener<String> completionListener,
```

Comment (andrross) [reply server/src/main/java/org/opensearch/common/blobstore/stream/read/listener/ReadContextListener.java:105]: This was intentional. The semantics of a GroupedActionListener requires that it be notified an exact number of times (success or failure). There's a fair bit of logic here to accomplish that so I thought the specific type helped convey that intent.

File: plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/S3BlobContainer.java (hunk 2 above)
Comment (andrross) [reply plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/S3BlobContainer.java:258]: > Curious what is the throughput 20 streams can drive?

In my test on a `r7a.xlarge` node I was able to get an effective throughput of about 333 MiB/s.

I'm adding a new dynamic recovery setting called `indices.recovery.max_concurrent_remote_store_streams` for this parameter. The existing `cluster.routing.allocation.node_concurrent_recoveries` and `indices.recovery.max_bytes_per_sec` settings are being honored here as well.

File: server/src/main/java/org/opensearch/common/blobstore/stream/read/listener/ReadContextListener.java (hunk 1 above)
Comment (andrross) [reply server/src/main/java/org/opensearch/common/blobstore/stream/read/listener/ReadContextListener.java]: Added a new dynamic setting (`indices.recovery.max_concurrent_remote_store_streams`). Unfortunately it was not super simple to wire in and I'm not totally happy with how I did it, but this seems like the best way absent a much larger refactoring.

**Review by kotwanikunal — APPROVED** [review]

**Comment by kotwanikunal** [comment]
#9115 

> ### Gradle Check (Jenkins) Run Completed with:
> * **RESULT:** FAILURE ❌
> * **URL:** https://build.ci.opensearch.org/job/gradle-check/26967/
> * **CommitID:** [a9b8b63](https://github.com/opensearch-project/OpenSearch/commit/a9b8b6314a1769c909913444a80abf24dcb0b175)
>   Please examine the workflow log, locate, and copy-paste the failure(s) below, then iterate to green.
>   Is the failure [a flaky test](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) unrelated to your change?

**Comment by andrross** [comment]
I'm going to merge this to unblock subsequent PRs, but will follow up with @Bukhtawar and @gbbafna and address any comments or concerns.

**Pull request merged**